mvebu: add support for IIJ SA-W2

musashino205 opened the pull request:
This PR adds seil-fw mtdsplit driver and support for IIJ SA-W2 appliance.
This PR is a successor of #12021, with the replaced parser.

musashino205 committed:
mvebu: add support for IIJ SA-W2

Internet Initiative Japan Inc. (IIJ) SA-W2 is a network appliance with
11ac (Wi-Fi 5) wlan, based on 88F6810.

Specification:

- SoC         : Marvell Armada 380 88F6810
- RAM         : DDR3 256 MiB (Micron MT41K64M16TW-107:J x2)
- Flash       : SPI-NOR 32 MiB (Winbond W25Q256JVFIQ)
- WLAN        : 2.4/5 GHz, Mini PCI-E
  - 2.4 GHz   : Silex SX-PCEGN (Atheros AR9287 (2T2R))
  - 5 GHz     : Silex SX-PCEAC (Qualcomm Atheros QCA9880 (3T3R))
- Ethernet    : 10/100/1000 Mbps x5
  - Switch    : Marvell 88E6172
- LEDs/Keys   : 12x/1x
- UART        : "CONSOLE" port (RJ-45, RS-232C)
  - settings  : 115200n8
  - assignment: 1:NC,  2:NC,  3:TXD, 4:GND,
                5:GND, 6:RXD, 7:NC,  8:NC
  - note      : compatible with Cisco console cable
- Power       : DC Input or PoE
  - DC Input  : 12 VDC, 3 A
  - PoE       : 802.3af
    - module  : Silvertel Ag9712-2BR
    - note    : USB ports shouldn't be used when powered by PoE
- Bootloader  : PMON2000 based
- Stock       : NetBSD based

Flash instruction using sysupgrade image:

1. Prepare TFTP server with IP address 192.168.0.10 and put sysupgrade
   image to TFTP directory
2. Connect PC to "GE0/PoE" port on SA-W2
3. Power on SA-W2, interrupt count-down by Esc and enter to bootloader
   CLI
4. Set IP address of the device

   address 192.168.0.1

5. Download sysupgrade image and flash to storage

   tftpload 192.168.0.10 <image name>
   firmwrite

   example:

   #tftpload 192.168.0.10 openwrt-mvebu-cortexa9-iij_sa-w2-squashfs-sysupgrade.bin
   Loading openwrt-mvebu-cortexa9-iij_sa-w2-squashfs-sysupgrade.bin
   loaded 8127268 byte(s)
   #firmwrite
   Erasing FLASH block  32 Done 0x00200000.
   Erasing FLASH block  33 Done 0x00210000.
   ...
   Erasing FLASH block 155 Done 0x009b0000.
   Erasing FLASH block 156 Done 0x009c0000.
   Programming FLASH. Done.
   Verifying FLASH. No Errors found.

6. Check the flashed firmware

   firmcheck

   example:

   #firmcheck
   [Normal firmware]
   ident: 'SEIL2015'
   copyright: 'ARM OpenWrt Linux-5.15.93'
   version format: 1
   version major: 9
   version minor: 99
   version release: 'r22060+36-5163bb5e54'
   body size: 3578524
   checksum: 0x8a083cb8

   [Rescue firmware]
   ident: 'SEIL2015'
   copyright: 'Copyright (c) 2017 Internet Initiative Japan Inc. All rights reserved.'
   version format: 1
   version major: 3
   version minor: 70
   version release: 'Release'
   body size: 10152458
   checksum: 0x8f9518c2

7. Boot with the flashed firmware

   boot

Note:

- The bootloader on this device is not U-Boot and it's environment space
  ("bootloader-env") has no compatibility with U-Boot tools.

- eth1 is connected to port6 of 88E6172 switch, but multi-cpu port can't
  be handled on Linux Kernel and not defined.

- Powering by PoE hasn't been tested yet.

- This device has 2x OS images on flash and they can be switched by
  setting "BOOTDEV" variable on bootloader CLI.
  That variable supports the following values:

  - "flash" : primary image on flash ("firmware")
  - "rescue": secondary image on flash ("rescue")
  - "usb"   : usb storage (broken?)
  - "lan0/1": network

  command to set:

  set BOOTDEV=<dev>

  example:

    set BOOTDEV=rescue

  This commit also supports booting from secondary partition.

- To execute initramfs image on bootloader CLI, use "go" command.
  ("go" command is not listed on the output of "help", but available)

  example (download and execute):

    address 192.168.0.1
    tftpload 192.168.0.10 openwrt-mvebu-cortexa9-iij_sa-w2-initramfs-kernel.bin
    go

MAC addresses:

LAN    : 00:E0:4D:xx:xx:19 (none)
WAN    : 00:E0:4D:xx:xx:18 (board_info, 0x6 (hex))
2.4 GHz: 84:25:3F:xx:xx:xx (Mini PCI-E card)
5 GHz  : 84:25:3F:xx:xx:xx (Mini PCI-E card)

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
kernel: add seil-fw mtdsplit driver for IIJ SEIL devices

This mtdsplit parser driver parses firmware partition on Internet
Initiative Japan Inc. (IIJ) SEIL series devices.

Structure of header:

  0x0 - 0x7 : Identifier            (hex)
  0x8 - 0x57: Copyright             (ascii)
 0x58 - 0x5b: Data CRC              (hex)
 0x5c - 0x5f: Image Format Version  (hex)
 0x60 - 0x63: Image Major Version   (hex)
 0x64 - 0x67: Image Minor Version   (hex)
 0x68 - 0x87: Image Release Version (ascii)
 0x88 - 0x8b: Xor value for Data?   (hex)
 0x8c - 0x8f: Data Length           (hex)

Properties:

- compatible      : "iij,seil-firmware"
- iij,seil-id     : ID of SEIL firmware for the device (8 bytes)
  - examples:
    - SA-W2       : <0x5345494c 0x32303135> ("SEIL2015")
    - SEIL/X1     : <0x5345494c 0x2F582020> ("SEIL/X  ")
- iij,bootdev-name: boot device name assigned to the partition
                    (optional)

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
mvebu: cortexa9: enable seil-fw mtdsplit driver

Enable seil-fw driver on mvebu/cortexa9 to use it on IIJ SA-W2.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

BKPepe commented:
Was the DTS sent to upstream as discussed in the previous PR?

musashino205 commented:
Yes, I sent.

BKPepe commented:
I found this link https://lore.kernel.org/lkml/64b5504d-c81d-f77e-da67-c6d9e76aedd6@gmail.com/ , did you already sent v2 or I missed that?  Would be nice to include the patch here instead of dts, so everyone knows that you are working on it or it is already included in the Linux kernel, so it can be removed in the future here.

musashino205 commented:
No, I didn't. I was preparing v2, but I was busy and I put it on hold for a while.

rmilecki commented on target/linux/mvebu/files/arch/arm/boot/dts/armada-380-iij-sa-w2.dts:
```diff
@@ -0,0 +1,382 @@
+
+				macaddr_bootenv_6: macaddr@6 {
+					reg = <0x6 0x6>;
+				};
```
Please use `nvmem-layout` + `fixed-layout` + `mac-base`. That way you don't need `mac-address-increment` (it's OpenWrt downstream binding that will get removed eventually).

Example: f1492fcc091c ("ramips: use fixed layout cell "mac-base" for Netgear R6220").

musashino205 commented:
UPDATE:

- converted NVMEM node to ```nvmem-layout```
- fixed wrong dt label of NVMEM MAC address
  - ```macaddr_bootenv_6``` ---> ```macaddr_bdinfo_6```
- added dual-boot support to mtdsplit parser and device support
- added ```fw_devlink=permissive``` parameter to bootargs
  - [to prevent looping detection of ethernet phys in Marvell 88E6172 switch](https://gist.github.com/musashino205/e057f05298d899332e215ab03422be58)
  - note: https://github.com/openwrt/openwrt/commit/5356462ce54734afd32fce83ad118c58cfeb2a55

hauke commented on target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_seil.c:
```diff
@@ -0,0 +1,204 @@
+	else if (ret < 0)
+		return false;
+
+	if (IS_ERR(env_mtd)) {
```
If you have to do an error check I prefer if you do the assignment of `env_mtd` in the line before. Something like this:
````
env_mtd = get_mtd_device_nm(LDR_ENV_PART_NAME);
if (IS_ERR(env_mtd)) {
....
````

musashino205 replied:
Indeed, I'll update.

hauke commented on target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_seil.c:
```diff
@@ -0,0 +1,204 @@
+		return -ENOMEM;
+
+	buflen = sizeof(struct seil_header);
+	buf = vmalloc(buflen);
```
`buf` is about 140 bytes big, I would prefer `kmalloc()` here.

musashino205 replied:
I used the same since mtdsplit_uimage is using ```vmalloc``` for header.
But indeed, the sizes of those headers are small enough, so ```kmalloc``` is better.

musashino205 replied:
I mage an additional change to read header directly without buf:
https://github.com/musashino-build/openwrt/commit/f1a387376dde862019ef21416e30db29cdb2d65a

Should this change also be applied?

hauke replied:
This looks fine. I think the `(void *)` is not needed.

musashino205 replied:
Thank you. I tried to removing ```(void *)```, but failed to compile due to ```incompatible pointer type```.

```
drivers/mtd/mtdsplit/mtdsplit_seil.c: In function 'mtdsplit_parse_seil_fw':
drivers/mtd/mtdsplit/mtdsplit_seil.c:128:52: error: passing argument 5 of 'mtd_read' from incompatible pointer type [-Werror=incompatible-pointer-types]
  128 |         ret = mtd_read(master, 0, hdrlen, &retlen, &header);
      |                                                    ^~~~~~~
      |                                                    |
      |                                                    struct seil_header *
In file included from drivers/mtd/mtdsplit/mtdsplit_seil.c:11:
./include/linux/mtd/mtd.h:508:22: note: expected 'u_char *' {aka 'unsigned char *'} but argument is of type 'struct seil_header *'
  508 |              u_char *buf);
      |              ~~~~~~~~^~~
```

I'll apply https://github.com/musashino-build/openwrt/commit/f1a387376dde862019ef21416e30db29cdb2d65a without removing ```(void *)``` and update this PR...

hauke replied:
ok makes sense. I thought it gets a `void *`, but it gets a `u_char *`. 
The cast is fine.

musashino205 commented:
UPDATE:

- DeviceTree:

  - fix label-mac-device
    - ```eth0``` is not enabled, use ```ge0``` on switch instead
  - add ```button-``` prefix to keys in gpio-keys

- mtdsplit_seil:

  - move ```get_mtd_device_nm()``` before error checking
  - use ```kmalloc``` instead of ```vmalloc``` for reading header
  - fix handling of return value on reading error of env_mtd
  - don't handle unused rootfs type

musashino205 commented:
UPDATE:

- mtdsplit_seil:
  - read seil header directly without buffer
